include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -450,13 +451,37 @@ namespace internal
     {
       Assert(n_rows == n_cols, ExcNotImplemented());
 
-      auto &&transpose_fill_nm = [](Number *           out,
-                                    const Number *     in,
-                                    const unsigned int n,
-                                    const unsigned int m) {
-        for (unsigned int mm = 0; mm < m; ++mm)
-          for (unsigned int nn = 0; nn < n; ++nn)
-            out[mm + nn * m] = *(in++);
+      std::vector<bool> constrained_dofs(n_rows, false);
+
+      for (unsigned int i = 0; i < n_rows; ++i)
+        {
+          if (mass_matrix[i + i * n_rows] == 0.0)
+            {
+              Assert(derivative_matrix[i + i * n_rows] == 0.0,
+                     ExcInternalError());
+
+              for (unsigned int j = 0; j < n_rows; ++j)
+                {
+                  Assert(mass_matrix[i + j * n_rows] == 0, ExcInternalError());
+                  Assert(mass_matrix[j + i * n_rows] == 0, ExcInternalError());
+                  Assert(derivative_matrix[i + j * n_rows] == 0,
+                         ExcInternalError());
+                  Assert(derivative_matrix[j + i * n_rows] == 0,
+                         ExcInternalError());
+                }
+
+              constrained_dofs[i] = true;
+            }
+        }
+
+      const auto transpose_fill_nm = [&constrained_dofs](Number *           out,
+                                                         const Number *     in,
+                                                         const unsigned int n,
+                                                         const unsigned int m) {
+        for (unsigned int mm = 0, c = 0; mm < m; ++mm)
+          for (unsigned int nn = 0; nn < n; ++nn, ++c)
+            out[mm + nn * m] =
+              (mm == nn && constrained_dofs[mm]) ? Number(1.0) : in[c];
```

Comment:
Much better.

Reply:
I also prefer the new variable `c` rather than pointer arithmetic. You could also have written `in[mm * n + nn]` in my opinion, but this is up to personal preference.